### ISSUE TITLE
batch raft call deadlock

### ISSUE THREAD

**Issue by BusyJay** [tikv/tikv#7493]
### What version of TiKV are you using?
```
$ tikv-server --version
TiKV
Release Version:   4.0.0-rc
Git Commit Hash:   f45d0c963df3ee4b1011caf5eb146cacd1fbbad8
Git Commit Branch: heads/v4.0.0-rc
UTC Build Time:    2020-04-08 11:18:15
Rust Version:      rustc 1.42.0-nightly (0de96d37f 2019-12-19)
Enable Features:   jemalloc portable protobuf-codec
Profile:           dist_release
```

### What operating system and CPU are you using?
Linux/PowerPC

### Steps to reproduce
Not sure how to reproduce

### What did happened?
90082 and 90078 are deadlock which blocks grpc thread can can prevent messages sent by raftstore hence causing infinite election.

Attachment is thread stacks caught in different time.

[gdb-1.txt](https://github.com/tikv/tikv/files/4486001/gdb-1.txt)
[gdb-2.txt](https://github.com/tikv/tikv/files/4486002/gdb-2.txt)
[gdb-3.txt](https://github.com/tikv/tikv/files/4486004/gdb-3.txt)

**Comment by BusyJay** [comment]
The lock maybe acquired by thread other than 90082 and 90078.

**Comment by siddontang** [comment]
have we found the problem yet?

**Comment by hunterlxt** [comment]
Sorry to late respond. The 3 attachment texts show that 2 thread stuck in the acquisition lock. And their thread stack shows that `BatchCommandsNotify::notify` has been called twice. In `notify(id)`, thread try to get the lock. In short, this issue is caused by self-locking.

`poll_future_notify(notify, id)` want to notify again, because in the related future task, `send_and_notify` will be called and the task will be called by `notify(id)` again. 

with help of @BusyJay , I'm trying to add fail test into Tikv and try to fix it.

**Comment by BusyJay** [comment]
The issue is introduced by https://github.com/tikv/tikv/pull/4473 and https://github.com/tikv/tikv/pull/5001. And https://github.com/pingcap/tidb/pull/14145 sets a 0 delay for empty message cause the dead lock. Asked @lysu that the change is only available in TiDB 4.0 and master, which explains why only 4.0 can have the problem.

A quick fix is set a non-zero delay for empty request. A permanent fix is avoid recursively notify in `BatchCommandsNotify`.
